File: src/backend/websocket/s2s_events.py
import json
from utils import get_agent_config

config = get_agent_config()

# Stringify inputSchema JSON definitions for tools
for tool in config['DEFAULT_TOOL_CONFIG']['tools']:
  schema = tool['toolSpec']['inputSchema']['json']
  tool['toolSpec']['inputSchema']['json'] = json.dumps(schema)

class S2sEvent:
  # Default configuration values
  DEFAULT_INFER_CONFIG = config['DEFAULT_INFER_CONFIG']
  DEFAULT_SYSTEM_PROMPT = config['DEFAULT_SYSTEM_PROMPT']
  DEFAULT_AUDIO_INPUT_CONFIG = config['DEFAULT_AUDIO_INPUT_CONFIG']
  DEFAULT_AUDIO_OUTPUT_CONFIG = config['DEFAULT_AUDIO_OUTPUT_CONFIG']
  DEFAULT_TOOL_CONFIG = config['DEFAULT_TOOL_CONFIG']

  @staticmethod
  def session_start(inference_config=DEFAULT_INFER_CONFIG): 
    return {"event":{"sessionStart":{"inferenceConfiguration":inference_config}}}

  @staticmethod
  def prompt_start(prompt_name, 
                   audio_output_config=DEFAULT_AUDIO_OUTPUT_CONFIG, 
                   tool_config=DEFAULT_TOOL_CONFIG):
    return {
          "event": {
            "promptStart": {
              "promptName": prompt_name,
              "textOutputConfiguration": {
                "mediaType": "text/plain"
              },
              "audioOutputConfiguration": audio_output_config,
              "toolUseOutputConfiguration": {
                "mediaType": "application/json"
              },
              "toolConfiguration": tool_config
            }
          }
        }

  @staticmethod
  def content_start_text(prompt_name, content_name):
    return {
        "event":{
        "contentStart":{
          "promptName":prompt_name,
          "contentName":content_name,
          "type":"TEXT",
          "interactive":False,
          "role": "SYSTEM",
          "textInputConfiguration":{
            "mediaType":"text/plain"
            }
          }
        }
      }
    
  @staticmethod
  def text_input(prompt_name, content_name, system_prompt=DEFAULT_SYSTEM_PROMPT):
    return {
      "event":{
        "textInput":{
          "promptName":prompt_name,
          "contentName":content_name,
          "content":system_prompt,
        }
      }
    }
  
  @staticmethod
  def content_end(prompt_name, content_name):
    return {
      "event":{
        "contentEnd":{
          "promptName":prompt_name,
          "contentName":content_name
        }
      }
    }

  @staticmethod
  def content_start_audio(prompt_name, content_name, audio_input_config=DEFAULT_AUDIO_INPUT_CONFIG):
    return {
      "event":{
        "contentStart":{
          "promptName":prompt_name,
          "contentName":content_name,
          "type":"AUDIO",
          "interactive":True,
          "audioInputConfiguration":audio_input_config
        }
      }
    }
    
  @staticmethod
  def audio_input(prompt_name, content_name, content):
    return {
      "event": {
        "audioInput": {
          "promptName": prompt_name,
          "contentName": content_name,
          "content": content,
        }
      }
    }
  
  @staticmethod
  def content_start_tool(prompt_name, content_name, tool_use_id):
    return {
        "event": {
          "contentStart": {
            "promptName": prompt_name,
            "contentName": content_name,
            "interactive": False,
            "type": "TOOL",
            "role": "TOOL",
            "toolResultInputConfiguration": {
              "toolUseId": tool_use_id,
              "type": "TEXT",
              "textInputConfiguration": {
                "mediaType": "text/plain"
              }
            }
          }
        }
      }
  
  @staticmethod
  def text_input_tool(prompt_name, content_name, content):
    return {
      "event": {
        "toolResult": {
          "promptName": prompt_name,
          "contentName": content_name,
          "content": content,
          #"role": "TOOL"
        }
      }
    }
  
  @staticmethod
  def prompt_end(prompt_name):
    return {
      "event": {
        "promptEnd": {
          "promptName": prompt_name
        }
      }
    }
  
  @staticmethod
  def session_end():
    return  {
      "event": {
        "sessionEnd": {}
      }
    }

File: src/frontend/src/helper/s2sEvents.js
import { DEFAULT_AUDIO_INPUT_CONFIG, DEFAULT_AUDIO_OUTPUT_CONFIG, DEFAULT_INFER_CONFIG, DEFAULT_SYSTEM_PROMPT, DEFAULT_TOOL_CONFIG } from '../agent/config.json'

// Stringify inputSchema JSON definitions for tools
DEFAULT_TOOL_CONFIG.tools.forEach(tool => {
  const schema = tool.toolSpec.inputSchema.json;
  tool.toolSpec.inputSchema.json = JSON.stringify(schema);
});

class S2sEvent {
  static DEFAULT_INFER_CONFIG = DEFAULT_INFER_CONFIG;
  static DEFAULT_SYSTEM_PROMPT = DEFAULT_SYSTEM_PROMPT;
  static DEFAULT_AUDIO_INPUT_CONFIG = DEFAULT_AUDIO_INPUT_CONFIG;
  static DEFAULT_AUDIO_OUTPUT_CONFIG = DEFAULT_AUDIO_OUTPUT_CONFIG;
  static DEFAULT_TOOL_CONFIG = DEFAULT_TOOL_CONFIG;

  static sessionStart(inferenceConfig = S2sEvent.DEFAULT_INFER_CONFIG, turnSensitivity = "LOW") {
    return {
      event: {
        sessionStart: {
          inferenceConfiguration: inferenceConfig
        },
        // turnDetectionConfiguration: {
        //   endpointingSensitivity: turnSensitivity
        // }
      }
    };
  }

  static promptStart(promptName, audioOutputConfig = S2sEvent.DEFAULT_AUDIO_OUTPUT_CONFIG, toolConfig = S2sEvent.DEFAULT_TOOL_CONFIG) {
    return {
      "event": {
        "promptStart": {
          "promptName": promptName,
          "textOutputConfiguration": {
            "mediaType": "text/plain"
          },
          "audioOutputConfiguration": audioOutputConfig,

          "toolUseOutputConfiguration": {
            "mediaType": "application/json"
          },
          "toolConfiguration": toolConfig
        }
      }
    }
  }

  static contentStartText(promptName, contentName, role = "SYSTEM", interactive = false) {
    return {
      "event": {
        "contentStart": {
          "promptName": promptName,
          "contentName": contentName,
          "type": "TEXT",
          "interactive": interactive,
          "role": role,
          "textInputConfiguration": {
            "mediaType": "text/plain"
          }
        }
      }
    }
  }

  static textInput(promptName, contentName, systemPrompt = S2sEvent.DEFAULT_SYSTEM_PROMPT) {
    var evt = {
      "event": {
        "textInput": {
          "promptName": promptName,
          "contentName": contentName,
          "content": systemPrompt
        }
      }
    }
    return evt;
  }

  static contentEnd(promptName, contentName) {
    return {
      "event": {
        "contentEnd": {
          "promptName": promptName,
          "contentName": contentName
        }
      }
    }
  }

  static contentStartAudio(promptName, contentName, audioInputConfig = S2sEvent.DEFAULT_AUDIO_INPUT_CONFIG) {
      return {
        "event": {
          "contentStart": {
            "promptName": promptName,
            "contentName": contentName,
            "type": "AUDIO",
            "interactive": true,
            "audioInputConfiguration": audioInputConfig
          }
        }
      }
    }

  static audioInput(promptName, contentName, content) {
    return {
      event: {
        audioInput: {
          promptName,
          contentName,
          content,
        }
      }
    };
  }

  static contentStartTool(promptName, contentName, toolUseId) {
      return {
        event: {
          contentStart: {
            promptName,
            contentName,
            interactive: false,
            type: "TOOL",
            role: "TOOL",
            toolResultInputConfiguration: {
              toolUseId,
              type: "TEXT",
              textInputConfiguration: { mediaType: "text/plain" }
            }
          }
        }
      };
    }

  static textInputTool(promptName, contentName, content) {
      return {
        event: {
          toolResult: {
            promptName,
            contentName,
            content,
          }
        }
      };
    }

  static promptEnd(promptName) {
    return {
      event: {
        promptEnd: {
          promptName
        }
      }
    };
  }

  static sessionEnd() {
    return { event: { sessionEnd: {} } };
  }
}
export default S2sEvent;